services/kubelet.go

```diff
@@ -0,0 +1,114 @@
+package services
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/go-connections/nat"
+	"github.com/rancher/rke/hosts"
+)
+
+type Kubelet struct {
+	Version string `yaml:"version"`
+	Image   string `yaml:"image"`
+	ClusterDomain	string `yaml:"cluster_domain"`
+	InfraContainerImage string `yaml:"infra_container_image"`
+}
+
+func runKubelet(host hosts.Host, masterHost hosts.Host, kubeletService Kubelet, isMaster bool) error {
+	isRunning, err := IsContainerRunning(host, KubeletContainerName)
```

Comment:
It feels like the code IsContainerRunning/PullImage/doRun is being repeated in each component. See how you can make it more generic and reusable (including the logs).

Reply:
Only building the config is really component specific